### ISSUE TITLE
Links ambiguous to AT users only can pass 2.4.4 and 2.4.9 

### ISSUE THREAD

**Issue by dd8** [w3c/wcag#1202]
> Success Criterion 2.4.4 Link Purpose (In Context) (Level A): The purpose of each link can be determined from the link text alone or from the link text together with its programmatically determined link context, except where the purpose of the link would be ambiguous to users in general.
https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context.html

> Success Criterion 2.4.9 Link Purpose (Link Only) (Level AAA): A mechanism is available to allow the purpose of each link to be identified from link text alone, except where the purpose of the link would be ambiguous to users in general.
https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-link-only

Since `aria-label` and `aria-labelledby` completely replace the visible link text you can get this situation:

```
<a href="http://facebook.com" aria-label="Follow us">Facebook</a>
<a href="http://twitter.com" aria-label="Follow us">Twitter</a>
```

The purpose of both links can be determined from the link text alone, but in AT both links are voiced as "Follow us"

**Comment by dd8** [comment]
Also in the understanding document:

> Identifying the purpose of a link using link text combined with programmatically determined link context using one of the following techniques:
ARIA7: Using aria-labelledby for link purpose
ARIA8: Using aria-label for link purpose

This seems to assume the link text is merged with `aria-labe`l and `aria-labelledby`

**Comment by JAWS-test** [comment]
> The purpose of both links can be determined from the link text alone, but in AT both links are voiced as "Follow us"

I don't see a problem because although 2.4.4 and 2.4.9 are met (visible link text is meaningful), criteria 1.3.1 and 2.5.3 are not met for AT users. Perhaps 2.4.4 and 2.4.9. should emphasize more clearly that 2.4.9. is about the visible link text and 2.4.4. is about the visible link text + its context (which must be both visible and AT context).

This is analogous to 2.4.6, which is also about the visible label and heading.

**Comment by JAWS-test** [comment]
> This seems to assume the link text is merged with aria-label and aria-labelledby

I don't see a problem here either, since there is a clear warning e.g. with ARIA8:

> Per the Accessible Name and Description Computation and the HTML to Platform Accessibility APIs Implementation Guide, the aria-label text will override the text supplied within the link. As such the text supplied will be used instead of the link text by AT. Due to this it is recommended to start the text used in aria-label with the text used within the link. This will allow consistent communication between users

It could be added in the passage that, according to 2.5.3, this is not only recommended but necessary

**Comment by dd8** [comment]
You can get round 2.5.3 by doing this:

```
<a href="http://facebook.com" aria-label="Follow us on Facebook or Twitter">Facebook</a>
<a href="http://twitter.com" aria-label="Follow us on Facebook or Twitter">Twitter</a>
```

So is the example above a normative fail of:

> Success Criterion 1.3.1 Info and Relationships (Level A): Information, structure, and relationships conveyed through presentation can be programmatically determined or are available in text.

**Comment by JAWS-test** [comment]
> You can argue here that although there is a programmatically determined aria-label, the information is also available in text (the link text) since the normative text uses "or"

Yes, I understand, there seems to be a gap in the WCAG. I would be in favour of making it clear in 1.3.1 that OR refers to the fact that information can also be transmitted in text form if it is not transmitted in the way of "programmatically determined"

**Comment by mraccess77** [comment]
@dd8 if you look at the examples 2 and 3 in ARIA 8 you will find that they aria-labelledby actually references both the inner text of the link and the neighboring contextual text -- so it is not implying that they are automatically combined - in fact it is showing that you need to include both in the reference.

**Comment by dd8** [comment]
The use of "link text" rather than accessible name looks especially odd in 2.4.4 which refers to programmatically determined link context. The Benefits section in the understanding document lists a mix of user types, some of who use the accessible name and some who use the visible label:
https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context.html

Looking at earlier WCAG 2.0 drafts the language implies it was originally intended for AT only:
> 2.4.4 Each link is programmatically associated with text from which its purpose can be determined
https://www.w3.org/TR/2006/WD-WCAG20-20060427/guidelines.html#navigation-mechanisms

The main part of the current language for 2.4.4 was introduced in 2007: 
> 2.4.4 Link Purpose (Context): The purpose of each link can be determined from the link text and its programmatically determined link context. 
https://www.w3.org/TR/2007/WD-WCAG20-20070517/#navigation-mechanisms

At the time the SC was written there was no ARIA and no accessible name calculation, and there wasn't a difference between accessible name and visible label, so I'm guessing the language was modified between 2006 and 2007 to benefit additional user groups.

Fast forward 13 years, there's now a clear distinction between between accessible name and visible label, which makes the 2.4.4 normative text problematic

**Comment by patrickhlauke** [comment]
Coming in late on this, would an acceptable interpretation (which does not require a change of normative language) be that for AT users, in the examples provided

```
<a href="http://facebook.com" aria-label="Follow us on Facebook or Twitter">Facebook</a>
<a href="http://twitter.com" aria-label="Follow us on Facebook or Twitter">Twitter</a>
```

that the "link text" for AT users is in effect the `aria-label`, so that you need to consider THAT when determining if the link passes or fails 2.4.4 / 2.4.9? i.e. this may only need a patch in understanding to explicitly clarify that for AT users, what counts is the *accessible name + accessible description* of the link, which may differ from the visible text?

**Comment by patrickhlauke** [comment]
Made a first stab at this https://github.com/w3c/wcag/pull/1570

**Comment by patrickhlauke** [comment]
> Looking at the benefits sections I think there's a mix of accessible name and visible label implied due to the different groups this SC benefits

as for 2.5.3 Label in Name the accessible name needs to at least contain the visible label, this is probably fine/not a problem?

**Comment by patrickhlauke** [comment]
> Would it be clearer to make "link text" a definition in the understanding document linking to the note, similar to the "purpose of each link" and "ambiguous to users in general" definitions

that would require adding "link text" as a normative definition in core WCAG itself and then making "link text" in the normative wording of the SC itself a link (at least that's the way those definitions currently work based on how understanding documents are built behind the scenes ... only normative terms which are referenced in the normative SC text get added to that part of an understanding document).

it's certainly feasible, but more technically involved and, as it's a normative change, probably take longer to be accepted / not possible to do retrospectively for 2.0/2.1.

**Comment by stes-acc** [comment]
I don't get why this discussion and edits like [(https://github.com/w3c/wcag/pull/1570] is/are not done on a broader scope.

Using aria-label / aria-description in concurrency to ANY visible text for a control (may it be a text node, a title attribute exposed in tooltip etc.) is always overriding the speech output for screen reader users for reasons described in [https://w3c.github.io/accname/]. The link text overriding is t JUST ONE example for this. Wouldn't it be better to create a designated separate page within WCAG space for advice and use of best practices/techniques all related requirements can refer to?

**Comment by patrickhlauke** [comment]
@stes-acc i took the liberty of editing your comment to correct the broken markdown you had for the links

**Comment by bruce-usab** [comment]
```
<a href="http://facebook.com" aria-label="Follow us">Facebook</a>
<a href="http://twitter.com" aria-label="Follow us">Twitter</a>
```
> The purpose of both links can be determined from the link text alone, but in AT both links are voiced as "Follow us"

FWIW, as an auditor, I think this falls under [Failure of Success Criterion 2.4.9 due to using a non-specific link such as "click here" or "more" without a mechanism to change the link text to specific text.](https://www.w3.org/WAI/WCAG22/Techniques/failures/F84)  Does anyone disagree?

My apologies if someone has already made this specific point.  I didn’t read the entire conversation closely and I see that @patrickhlauke is addressing!

**Comment by patrickhlauke** [comment]
```
<a href="http://facebook.com" aria-label="Follow us">Facebook</a>
<a href="http://twitter.com" aria-label="Follow us">Twitter</a>
```

noting in passing that those links would fail [2.5.3 Label in Name](https://www.w3.org/TR/WCAG22/#label-in-name)

**Comment by m13v** [comment]
the AX layer is the right place to catch this. When AT consumes a link, it reads the accessible name, which aria-label fully replaces. So 2.4.4 evaluated against rendered text passes (Facebook, Twitter visible labels are distinct), but evaluated against accessible name fails (both announce 'Follow us'). The fix is shifting the SC to operate on the accessible name rather than 'link text', which catches the duplicate aria-label case without breaking anything currently compliant. Same gap shows up in macOS/iOS AX trees, the visible label and the AX name diverge constantly.
